File: java/src/iBinaryTrees/PrintAllPaths.java
package iBinaryTrees;

public class PrintAllPaths {

	public void printPaths() {
		  int[] path = new int[1000];
		  BinaryTree root = new BinaryTree(1);
		  printPaths(root, path, 0);
	}
	
	public void printPaths(BinaryTree node, int[] path, int pathLen){
		if(node == null) return;
		path[pathLen] = node.getData();
		pathLen++;
		
		if(node.getLeft() == null && node.getRight() == null){
			printArray(path, pathLen);
		}
		else
		{
			printPaths(node.getLeft(), path, pathLen);
			printPaths(node.getRight(), path, pathLen);
		}
		
	}
	
	private void printArray(int[] ints, int len){
		int i;
		for(i = 0; i<len; i++){
			System.out.println(ints[i] + " ");
		}
		System.out.println();
	}
}

File: python/trees/print_all_paths.py


class Node:
    def __init__(self, data):
        self.data = data
        self.left, self.right = None, None


class BinaryTree(object):

    def __init__(self):
        self.root = None

    def add_node(self, data):
        return Node(data)

    def insert(self, root, data):
        if root is None:
            return self.add_node(data)
        else:
            if data <= root.data:
                root.left = self.insert(root.left, data)
            else:
                root.right = self.insert(root.right, data)
            return root

    def print_paths(self, root):
        self.print_paths_helper(root, [])

    def print_paths_helper(self, root, paths):
        if not root:
            return
        paths.append(root.data)
        if root.left is None and root.right is None:
            print "Printing path"
            print "->".join(str(p) for p in paths)
        else:
            self.print_paths_helper(root.left, list(paths))
            self.print_paths_helper(root.right, list(paths))


class BinaryTreeRunner(object):

    def __init__(self):
        self.btree = BinaryTree()
        self.btree.root = self.btree.insert(None, 5)
        root_3 = self.btree.insert(self.btree.root, 3)
        root_7 = self.btree.insert(self.btree.root, 7)
        self.btree.insert(root_3, 2)
        self.btree.insert(root_3, 4)
        self.btree.insert(root_7, 6)
        self.btree.insert(root_7, 8)

    def print_all_paths(self):
        self.btree.print_paths(self.btree.root)

    @staticmethod
    def main():
        BinaryTreeRunner().print_all_paths()


BinaryTreeRunner().main()




